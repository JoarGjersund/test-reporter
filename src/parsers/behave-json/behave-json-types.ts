export interface BehaveJson {
  stats: BehaveJsonStats
  passes: BehaveJsonTest[]
  pending: BehaveJsonTest[]
  failures: BehaveJsonTest[]
}

export interface BehaveJsonStats {
  duration: number
}

export interface BehaveJsonTest {
  title: string
  fullTitle: string
  file: string
  duration?: number
  err: BehaveJsonTestError
}

export interface BehaveJsonTestError {
  stack?: string
  message?: string
}
