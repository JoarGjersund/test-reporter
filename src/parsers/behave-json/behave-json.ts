import { ParseOptions, TestParser } from '../../test-parser'
import {
  TestCaseError,
  TestCaseResult,
  TestExecutionResult,
  TestGroupResult,
  TestRunResult,
  TestSuiteResult
} from '../../test-results'
import { getExceptionSource } from '../../utils/node-utils'
import { getBasePath, normalizeFilePath } from '../../utils/path-utils'
import { BehaveJson, BehaveJsonTest } from './behave-json-types'

export class BehaveJsonParser implements TestParser {
  assumedWorkDir: string | undefined

  constructor(readonly options: ParseOptions) { }

  async parse(path: string, content: string): Promise<TestRunResult> {
    const behave = this.getBehaveJson(path, content)
    const result = this.getTestRunResult(path, behave)
    result.sort(true)
    return Promise.resolve(result)
  }

  private getBehaveJson(path: string, content: string): BehaveJson {
    try {
      return JSON.parse(content)
    } catch (e) {
      throw new Error(`Invalid JSON at ${path}\n\n${e}`)
    }
  }

  // New parser for the provided JSON format
  private getTestRunResult(resultsPath: string, behave: any): TestRunResult {
    const suites: TestSuiteResult[] = [];
    let totalDuration = 0;

    if (!behave.features || !Array.isArray(behave.features)) {
      return new TestRunResult(resultsPath, [], 0);
    }

    for (const feature of behave.features) {
      const suiteName = feature.name || feature.filename || "Feature";
      const suitePath = this.getRelativePath(feature.filename || suiteName);
      const suite = new TestSuiteResult(suitePath, []);

      if (feature.scenarios && Array.isArray(feature.scenarios)) {
        for (const scenario of feature.scenarios) {
          // Group by scenario name
          let group = suite.groups.find(g => g.name === scenario.name);
          if (!group) {
            group = new TestGroupResult(scenario.name, []);
            suite.groups.push(group);
          }

          // Determine result status
          let status: TestExecutionResult = "skipped";
          if (scenario.status === "passed" || scenario.status === "success") status = "success";
          else if (scenario.status === "failed" || scenario.status === "error") status = "failed";
          else if (scenario.status === "untested" || scenario.status === "skipped") status = "skipped";

          // Aggregate step errors/details
          let error: TestCaseError | undefined = undefined;
          let stepDetails: string[] = [];
          if (scenario.steps && Array.isArray(scenario.steps)) {
            for (const step of scenario.steps) {
              // If any step is failed, mark scenario as failed
              if (step.status === "failed" || step.status === "error") status = "failed";
              // Collect step details
              stepDetails.push(`${step.step_type}: ${step.name} [${step.status}]`);
              if (step.status === "failed" && !error) {
                error = {
                  path: feature.filename,
                  line: undefined,
                  message: step.name,
                  details: step.text || ""
                };
              }
            }
          }

          // Compose test case
          // Attach step details to error.details if error exists, otherwise skip
          if (error) {
            error.details = (error.details ? error.details + "\n" : "") + stepDetails.join("\n");
          }
          const testCase = new TestCaseResult(
            scenario.name,
            status,
            scenario.duration || 0,
            error
          );
          group.tests.push(testCase);
          totalDuration += scenario.duration || 0;
        }
      }
      suites.push(suite);
    }
    return new TestRunResult(resultsPath, suites, totalDuration);
  }

  private processTest(suite: TestSuiteResult, test: BehaveJsonTest, result: TestExecutionResult): void {
    const groupName =
      test.fullTitle !== test.title
        ? test.fullTitle.substr(0, test.fullTitle.length - test.title.length).trimEnd()
        : null

    let group = suite.groups.find(grp => grp.name === groupName)
    if (group === undefined) {
      group = new TestGroupResult(groupName, [])
      suite.groups.push(group)
    }

    const error = this.getTestCaseError(test)
    const testCase = new TestCaseResult(test.title, result, test.duration ?? 0, error)
    group.tests.push(testCase)
  }

  private getTestCaseError(test: BehaveJsonTest): TestCaseError | undefined {
    const details = test.err.stack
    const message = test.err.message
    if (details === undefined) {
      return undefined
    }

    let path
    let line

    const src = getExceptionSource(details, this.options.trackedFiles, file => this.getRelativePath(file))
    if (src) {
      path = src.path
      line = src.line
    }

    return {
      path,
      line,
      message,
      details
    }
  }

  private getRelativePath(path: string): string {
    path = normalizeFilePath(path)
    const workDir = this.getWorkDir(path)
    if (workDir !== undefined && path.startsWith(workDir)) {
      path = path.substr(workDir.length)
    }
    return path
  }

  private getWorkDir(path: string): string | undefined {
    return (
      this.options.workDir ??
      this.assumedWorkDir ??
      (this.assumedWorkDir = getBasePath(path, this.options.trackedFiles))
    )
  }
}
