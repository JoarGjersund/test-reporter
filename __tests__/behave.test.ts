import * as fs from 'fs'
import * as path from 'path'

import {BehaveJsonParser} from '../src/parsers/behave-json/behave-json'
import {ParseOptions} from '../src/test-parser'
import {DEFAULT_OPTIONS, getReport} from '../src/report/get-report'
import {normalizeFilePath} from '../src/utils/path-utils'

describe('behave-json tests', () => {
  it('produces empty test run result when there are no features', async () => {
    const fixturePath = path.join(__dirname, 'fixtures', 'empty', 'behave.json')
    const filePath = normalizeFilePath(path.relative(__dirname, fixturePath))
    const fileContent = fs.readFileSync(fixturePath, {encoding: 'utf8'})

    const opts: ParseOptions = {
      parseErrors: true,
      trackedFiles: []
    }

    const parser = new BehaveJsonParser(opts)
    const result = await parser.parse(filePath, fileContent)
    expect(result.tests).toBe(0)
    expect(result.result).toBe('success')
  })

  it('report from behave.json test results matches snapshot', async () => {
    const fixturePath = path.join(__dirname, 'fixtures', 'behave.json')
    const outputPath = path.join(__dirname, '__outputs__', 'behave-json.md')
    const filePath = normalizeFilePath(path.relative(__dirname, fixturePath))
    const fileContent = fs.readFileSync(fixturePath, {encoding: 'utf8'})

    const opts: ParseOptions = {
      parseErrors: true,
      trackedFiles: []
    }

    const parser = new BehaveJsonParser(opts)
    const result = await parser.parse(filePath, fileContent)
    expect(result).toMatchSnapshot()

    const report = getReport([result])
    fs.mkdirSync(path.dirname(outputPath), {recursive: true})
    fs.writeFileSync(outputPath, report)
  })
})
