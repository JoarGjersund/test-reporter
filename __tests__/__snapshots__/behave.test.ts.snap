// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`behave-json tests report from behave.json test results matches snapshot 1`] = `
TestRunResult {
  "path": "fixtures/behave.json",
  "suites": [
    TestSuiteResult {
      "groups": [
        TestGroupResult {
          "name": "Example Scenario 1",
          "tests": [
            TestCaseResult {
              "error": undefined,
              "name": "Example Scenario 1",
              "result": "skipped",
              "time": 0,
            },
          ],
        },
      ],
      "name": "example.feature",
      "totalTime": undefined,
    },
  ],
  "totalTime": 0,
}
`;
